packages/deployer/src/compiler.ts

```diff
@@ -105,27 +115,48 @@ export class Compiler {
     public async compileAsync(): Promise<void> {
         await createDirIfDoesNotExistAsync(this._artifactsDir);
         await createDirIfDoesNotExistAsync(SOLC_BIN_DIR);
-        this._contractSources = await Compiler._getContractSourcesAsync(this._contractsDir);
+        this._contractSources = {};
+        const contractDirs = Array.from(this._contractDirs.values());
+        for (const contractDir of contractDirs) {
+            const sources = await Compiler._getContractSourcesAsync(contractDir.path, contractDir.path);
+            _.forIn(sources, (source, sourceFilePath) => {
+                const sourceFileId = constructUniqueSourceFileId(contractDir.namespace, sourceFilePath);
+
+                // Record the file's source and data
+                if (!_.isUndefined(this._contractSources[sourceFileId])) {
+                    throw new Error(`Found duplicate source files with ID '${sourceFileId}'`);
+                }
+                this._contractSources[sourceFileId] = source;
+
+                // Create a mapping between the contract id and its source file id
+                const contractId = constructContractId(contractDir.namespace, sourceFilePath);
+                if (!_.isUndefined(this._contractIds[contractId])) {
+                    throw new Error(`Found duplicate contract with ID '${contractId}'`);
+                }
+                this._contractIds[contractId] = sourceFileId;
```

Comment:
we may want to consider writing to the instance variables after this loop because it may terminate prematurely in an error case